File: babylon60.py
#!/usr/bin/env python3
# C5-REAL
import sys

def parse_b60_digit(token):
    if token == '-': 
        return 0
    tens = token.count('<')
    ones = token.count('Y') + token.count('v') + token.count('T')
    return tens * 10 + ones

def parse_b60_number(b60_str):
    inner = b60_str.strip('[]').strip()
    if not inner: 
        return 0
    places = inner.split()
    total = 0
    power = len(places) - 1
    for p in places:
        val = parse_b60_digit(p)
        total += val * (60 ** power)
        power -= 1
    return total

def format_b60(val):
    if val == 0: return "[-]"
    places = []
    while val > 0:
        places.append(val % 60)
        val //= 60
    places.reverse()
    
    out = []
    for p in places:
        if p == 0:
            out.append("-")
        else:
            tens = p // 10
            ones = p % 10
            out.append("<" * tens + "Y" * ones)
    return "[ " + " ".join(out) + " ]"

class Babylon60:
    def __init__(self):
        self.memory = {}

    def eval_expr(self, expr):
        if expr.startswith('['):
            return parse_b60_number(expr)
        else:
            return self.memory.get(expr, 0)

    def run(self, code):
        for line in code.split('\n'):
            line = line.split('#')[0].strip()
            if not line: continue
            
            tokens = []
            in_bracket = False
            cur = ""
            for char in line:
                if char == '[':
                    in_bracket = True
                    cur += char
                elif char == ']':
                    in_bracket = False
                    cur += char
                    tokens.append(cur.strip())
                    cur = ""
                elif char.isspace() and not in_bracket:
                    if cur:
                        tokens.append(cur)
                        cur = ""
                else:
                    cur += char
            if cur:
                tokens.append(cur)
                
            cmd = tokens[0]
            if cmd == 'DUB':       # Tablet initialization
                pass
            elif cmd == 'SAR':     # Print Decimal
                val = self.eval_expr(tokens[1])
                print(f"SAR (DEC): {val}")
            elif cmd == 'SAR.B60': # Print Base-60
                val = self.eval_expr(tokens[1])
                print(f"SAR (B60): {format_b60(val)}")
            elif cmd == 'NIG':     # Assign
                self.memory[tokens[1]] = self.eval_expr(tokens[2])
            elif cmd == 'DAH':     # Add
                self.memory[tokens[1]] = self.memory.get(tokens[1], 0) + self.eval_expr(tokens[2])
            elif cmd == 'LAL':     # Subtract
                self.memory[tokens[1]] = self.memory.get(tokens[1], 0) - self.eval_expr(tokens[2])
            elif cmd == 'ARA':     # Multiply
                self.memory[tokens[1]] = self.memory.get(tokens[1], 0) * self.eval_expr(tokens[2])
            elif cmd == 'BA':      # Divide
                divisor = self.eval_expr(tokens[2])
                if divisor != 0:
                    self.memory[tokens[1]] = self.memory.get(tokens[1], 0) // divisor
            else:
                print(f"Unknown command: {cmd}")

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Usage: python3 babylon60.py <script.b60>")
        sys.exit(1)
    
    with open(sys.argv[1], 'r') as f:
        code = f.read()
    
    interpreter = Babylon60()
    interpreter.run(code)

File: babylon60.rs
use std::collections::HashMap;
use std::env;
use std::fs;
use std::thread;
use std::time::Duration;

#[derive(Clone, Debug, PartialEq)]
enum B60Type {
    I64,
    TIME,
    F60,
    UNALLOCATED,
}

#[derive(Clone, Debug)]
struct Register {
    val: i64,
    typ: B60Type,
}

fn parse_b60_digit(token: &str) -> i64 {
    if token == "-" { return 0; }
    let tens = token.chars().filter(|&c| c == '<').count() as i64;
    let ones = token.chars().filter(|&c| c == 'Y' || c == 'v' || c == 'T').count() as i64;
    tens * 10 + ones
}

fn parse_b60_number(b60_str: &str) -> i64 {
    let inner = b60_str.trim_matches(|c| c == '[' || c == ']' || c == ' ');
    if inner.is_empty() { return 0; }
    let places: Vec<&str> = inner.split_whitespace().collect();
    let mut total = 0;
    let mut power = (places.len() - 1) as u32;
    for p in places {
        total += parse_b60_digit(p) * 60_i64.pow(power);
        if power > 0 { power -= 1; }
    }
    total
}

fn format_b60(mut val: i64) -> String {
    if val == 0 { return "[-]".to_string(); }
    let mut places = Vec::new();
    while val > 0 {
        places.push(val % 60);
        val /= 60;
    }
    places.reverse();
    let mut out = Vec::new();
    for p in places {
        if p == 0 {
            out.push("-".to_string());
        } else {
            let tens = p / 10;
            let ones = p % 10;
            let mut s = String::new();
            for _ in 0..tens { s.push('<'); }
            for _ in 0..ones { s.push('Y'); }
            out.push(s);
        }
    }
    format!("[ {} ]", out.join(" "))
}

fn get_reg_index(reg_str: &str) -> usize {
    if reg_str.starts_with('R') {
        reg_str[1..].parse().unwrap_or(0)
    } else {
        0
    }
}

fn eval_expr(expr: &str, registers: &[Register]) -> i64 {
    if expr.starts_with('[') {
        parse_b60_number(expr)
    } else if expr.starts_with('R') {
        let idx = get_reg_index(expr);
        registers[idx].val
    } else {
        0
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <script.b60>", args[0]);
        std::process::exit(1);
    }

    let code = fs::read_to_string(&args[1]).expect("Failed to read script");
    let lines: Vec<&str> = code.lines().map(|l| l.split('#').next().unwrap().trim()).collect();
    
    let mut registers = vec![Register { val: 0, typ: B60Type::UNALLOCATED }; 64];
    let mut labels: HashMap<String, usize> = HashMap::new();

    for (i, line) in lines.iter().enumerate() {
        if line.starts_with("MUB ") {
            let name = line.split_whitespace().nth(1).unwrap();
            labels.insert(name.to_string(), i);
        }
    }

    let mut pc = 0;
    while pc < lines.len() {
        let line = lines[pc];
        if line.is_empty() || line == "DUB" || line.starts_with("MUB ") {
            pc += 1;
            continue;
        }

        let mut tokens = Vec::new();
        let mut in_bracket = false;
        let mut in_string = false;
        let mut cur = String::new();
        for c in line.chars() {
            if c == '"' {
                in_string = !in_string;
                cur.push(c);
            } else if c == '[' && !in_string {
                in_bracket = true;
                cur.push(c);
            } else if c == ']' && !in_string {
                in_bracket = false;
                cur.push(c);
                tokens.push(cur.trim().to_string());
                cur.clear();
            } else if c.is_whitespace() && !in_bracket && !in_string {
                if !cur.is_empty() {
                    tokens.push(cur.clone());
                    cur.clear();
                }
            } else {
                cur.push(c);
            }
        }
        if !cur.is_empty() {
            tokens.push(cur);
        }
        if tokens.is_empty() { pc += 1; continue; }

        let cmd = tokens[0].as_str();
        match cmd {
            "ALLOC" => {
                let typ_str = &tokens[1];
                let idx = get_reg_index(&tokens[2]);
                registers[idx].typ = match typ_str.as_str() {
                    "TIME" => B60Type::TIME,
                    "F60" => B60Type::F60,
                    _ => B60Type::I64,
                };
            }
            "NIG" => {
                let idx = get_reg_index(&tokens[1]);
                registers[idx].val = eval_expr(&tokens[2], &registers);
            }
            "AFTER" => {
                let idx = get_reg_index(&tokens[1]);
                let secs = registers[idx].val as u64;
                println!("[MOSKV CLOCK] Sleeping for {} base units...", secs);
                thread::sleep(Duration::from_secs(secs));
                if let Some(&target) = labels.get(&tokens[2]) {
                    pc = target;
                    continue;
                }
            }
            "EXECUTE" => {
                let action = tokens[1].trim_matches('"');
                println!("[MOSKV LEDGER DISPATCH] ⚡ {} ", action);
            }
            "SAR" => {
                let val = eval_expr(&tokens[1], &registers);
                println!("SAR (DEC): {}", val);
            }
            "SAR.B60" => {
                let val = eval_expr(&tokens[1], &registers);
                println!("SAR (B60): {}", format_b60(val));
            }
            "HALT" => {
                break;
            }
            _ => {}
        }
        pc += 1;
    }
}
